### PR TITLE
feat: validate the public key embedded in the deposit script

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/515.

## Changes

* Add a `can_sign` column to the `deposit_signers` table.
* Rename the `signer_public_key` column to `signers_public_key`.
* Add a `can_sign_deposit_tx` query that checks for whether the signer is in the signing set associated with the aggregate key in the deposit script.
* Do the signing set check making a decision for whether the deposit is accepted.

## Testing Information

This PR adds an integration test for the query, another for whether we correctly say that we cannot sign for the transaction if we aren't part of the signing set, and modifies a few others that implicitly do the check for when we are in the signing set.

## Checklist:

- [ ] I have performed a self-review of my code

### CODE REVIEW

File: signer/src/transaction_signer.rs [hunk 1]
```diff
@@ -639,42 +640,52 @@ where
...
             output_index: request.output_index,
             accepted: is_accepted,
+            can_sign,
         };
 
         let signer_decision = model::DepositSigner {
             txid: request.txid,
             output_index: request.output_index,
-            signer_pub_key: self.signer_pub_key(),
+            signer_pub_key: signer_public_key,
```
Comment (cylewitruk) [inline signer/src/transaction_signer.rs:686]: Suggestion: Not related specifically to this line, but after reading through a number of files I'm starting to dislike `signers_public_keys` as we also have `signer_public_key` like here. I'd almost prefer that we called it `signer_set_public_key`. Verbose, but very clear... thoughts?

File: signer/src/transaction_signer.rs [hunk 2]
```diff
@@ -738,6 +749,36 @@ where
...
+        // return early.
+        let Some(client) = self.blocklist_checker.as_ref() else {
+            return Ok(true);
+        };
+
+        // We turn all the input scriptPubKeys into addresses and check
+        // those with the blocklist client.
+        let bitcoin_network = bitcoin::Network::from(self.context.config().signer.network);
+        let params = bitcoin_network.params();
+        let addresses = req
```
Comment (cylewitruk) [inline signer/src/transaction_signer.rs:763]: Question: This is the stuff that crashed before, right? When working towards the demo and trying to convert to an address the blocklist client would accept? Do we know if this code fixes that?

**Review by cylewitruk — COMMENTED** [review]
Got mostly through but need to afk for a little bit, posting my comments/questions so far

File: signer/src/transaction_signer.rs (hunk 2 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:763]: > This is the stuff that crashed before, right? When working towards the demo and trying to convert to an address the blocklist client would accept?

Yeah, the code would crashed here.

> Do we know if this code fixes that if so?

That bug was fixed by https://github.com/stacks-network/sbtc/pull/675. This change is just a reorganization.

File: signer/src/transaction_signer.rs (hunk 1 above)
Comment (cylewitruk) [reply signer/src/transaction_signer.rs:686]: Just this comment alone left me overly confused 😂

But yeah, naming for the different concepts which are explicit and clear is needed regardless, imo

File: signer/src/transaction_signer.rs (hunk 2 above)
Comment (cylewitruk) [reply signer/src/transaction_signer.rs:763]: Awesome, just wanted to double check👌

File: signer/src/transaction_signer.rs (hunk 1 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:686]: We spoke about this yesterday. Basically, we should stick to a common name for the aggregate key in the code base. I'll open a ticket for the refactor for the name change stuff.

File: signer/src/storage/postgres.rs [hunk 3]
```diff
@@ -735,6 +735,38 @@ impl super::DbRead for PgStore {
         .map_err(Error::SqlxQuery)
     }
 
+    async fn can_sign_deposit_tx(
```
Comment (cylewitruk) [inline signer/src/storage/postgres.rs:738]: I'm not suggesting we change this here and now, but just wanted to note that this is an example of where I probably would have split this up into 1) getting the valid keys, 2) getting the deposit request, and 3) performing the validation in Rust.

I think that would have let you perform more of the testing as simpler unit tests.

File: signer/src/transaction_signer.rs (hunk 1 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:686]: I opened https://github.com/stacks-network/sbtc/issues/729 for the name changes.

**Review by cylewitruk — APPROVED** [review]
Just had one comment, but otherwise I think this looks good, approving since that comment is more of a suggestion than anything.

File: signer/src/storage/postgres.rs (hunk 3 above)
Comment (djordon) [reply signer/src/storage/postgres.rs:738]: Very good point

Comment (djordon) [reply signer/src/storage/postgres.rs:738]: 🙏🏽

**Pull request merged**